Allow updating of trusted cluster role maps

EdwardDowling opened the pull request:
closes [10568](https://github.com/gravitational/teleport/issues/10568)

EdwardDowling committed:
Allow updating of trusted cluster role maps

hatched commented on lib/auth/trustedcluster.go:
```diff
@@ -67,8 +68,15 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.
 
 	// change state
+
+	if exists && !cmp.Equal(existingCluster.GetRoleMap(), trustedCluster.GetRoleMap()) {
```
Can you add a test for this case plz.

espadolini replied:
nit: we can remove `exists` with `existingCluster != nil`; the check for `exists` should also probably only explicitly ignore a `NotFound` rather than blanket swallow all errors.

hatched reviewed (commented):
Could you add a test

EdwardDowling committed:
Add test for role map changes

hatched reviewed (approved):
Thanks

r0mant commented:
@espadolini @jakule Could you guys review this once you get a chance?

jakule commented on api/types/trustedcluster.go:
```diff
@@ -250,11 +250,12 @@ func (c *TrustedClusterV2) CanChangeStateTo(t TrustedCluster) error {
 		return immutableFieldErr("roles")
 	}
+	roleMapUpdated := false
 	if !cmp.Equal(c.GetRoleMap(), t.GetRoleMap()) {
-		return immutableFieldErr("role_map")
+		roleMapUpdated = true
 	}
```
nit:
```go
roleMapUpdated := !cmp.Equal(c.GetRoleMap(), t.GetRoleMap())
```

jakule commented on lib/auth/trustedcluster.go:
```diff
@@ -67,8 +68,15 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.
+	}
 	switch {
-	case exists == true && enable == true:
+	case exists == true && enable == true && existingCluster.GetEnabled() != trustedCluster.GetEnabled():
```
Can you store the result in a variable so we don't need to replace this check over and over?

```go
clusterDiffState := existingCluster.GetEnabled() != trustedCluster.GetEnabled()
```

I'm also opened to better naming.

r0mant replied:
Why did this condition change? Wouldn't this prevent user from upserting the trusted cluster if it's already enabled?

Same for the "exists == true && enable == false" case below.

r0mant commented on lib/auth/trustedcluster.go:
```diff
@@ -67,8 +68,15 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.
+		if err := a.checkLocalRoles(ctx, trustedCluster.GetRoleMap()); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		a.updateUserCARoleMap(ctx, existingCluster, trustedCluster)
```
Returned error is ignored.

r0mant commented on lib/auth/trustedcluster.go:
```diff
@@ -322,6 +322,23 @@ func (a *Server) addCertAuthorities(trustedCluster types.TrustedCluster, remoteC
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if certAuthority.GetType() == types.UserCA {
```
Why is this check needed? Aren't you querying specifically `types.UserCA` above?

r0mant commented on lib/auth/trustedcluster.go:
```diff
@@ -97,10 +105,6 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.
 	case exists == false && enable == true:
 		log.Debugf("Creating enabled Trusted Cluster relationship.")
 
-		if err := a.checkLocalRoles(ctx, trustedCluster.GetRoleMap()); err != nil {
```
Why was this (and the one below) check removed?

r0mant reviewed (changes requested):
Have a few questions.

r0mant commented:
@EdwardDowling Once you've implemented the changes, let's run through the following test matrix to test all possible scenarios. For each of these cases verify connectivity to leaf (should be able connect when tc is enabled, should not be able to connect when it's disabled) + role mapping.

- [ ] Create a new enabled trusted cluster.
- [ ] Create a new disabled trusted cluster.
- [ ] Disable enabled trusted cluster.
  - [ ] Then disable it back.
- [ ] Enable disabled trusted cluster.
  - [ ] Then enabled it back.
- [ ] Update role map on the enabled trusted cluster.
  - [ ] Verify new role map kicks in.
- [ ] Update role map on the disabled trusted cluster. Enable it.
  - [ ] Verify new role map kicks in.
- [ ] Delete trusted cluster.
- [ ] Edge-case scenarios like if you provide invalid role map.
- [ ] Verify role mapping can be updated in Terraform now: https://github.com/gravitational/teleport-plugins/issues/622.
- [ ] Verify we emit audit event when you update role map.

espadolini commented on lib/auth/trustedcluster.go:
```diff
@@ -322,6 +322,23 @@ func (a *Server) addCertAuthorities(trustedCluster types.TrustedCluster, remoteC
+	if certAuthority.GetType() == types.UserCA {
+		certAuthority.SetRoleMap(trustedCluster.GetRoleMap())
+	}
+	if err := a.UpsertCertAuthority(certAuthority); err != nil {
```
This should be a CAS operation, as the root cluster might have rotated its own user CA between the `GetCertAuthority` and the `UpsertCertAuthority` in this function, causing the local cluster to essentially roll back the remote CA, leading to very confusing behavior (not super critical, as it would eventually resync, but still).

espadolini reviewed (changes requested):
As this is a new piece of functionality, I think it would be fair to restrict it to simpler situations (like disallowing also going from enabled to disabled at the same time, for instance).

EdwardDowling commented:
> As this is a new piece of functionality, I think it would be fair to restrict it to simpler situations (like disallowing also going from enabled to disabled at the same time, for instance).

@r0mant what do you think of this idea?

r0mant commented:
> > As this is a new piece of functionality, I think it would be fair to restrict it to simpler situations (like disallowing also going from enabled to disabled at the same time, for instance).
> 
> @r0mant what do you think of this idea?

Is there an issue with updating role map alongside other trusted cluster changes? I don't object in general as main use-case I'm pretty sure is just updating role map without changing the state. Just trying to understand if we're not gonna be introducing another "artificial" restriction without a good reason. Does that simplify implementation?

EdwardDowling commented:
> > > As this is a new piece of functionality, I think it would be fair to restrict it to simpler situations (like disallowing also going from enabled to disabled at the same time, for instance).
> > 
> > 
> > @r0mant what do you think of this idea?
> 
> Is there an issue with updating role map alongside other trusted cluster changes? I don't object in general as main use-case I'm pretty sure is just updating role map without changing the state. Just trying to understand if we're not gonna be introducing another "artificial" restriction without a good reason. Does that simplify implementation?

The implementation would probably be about the same complexity. If there is no other benefit to that restriction I will continue finishing this implementation without it.

EdwardDowling committed:
Add integration tests for trusted cluster role map changes

EdwardDowling committed:
Move UpdateUserCARolemap to local/trust

espadolini commented on lib/services/local/trust.go:
```diff
@@ -293,6 +293,50 @@ func (s *CA) GetCertAuthorities(ctx context.Context, caType types.CertAuthType,
+		return trace.Wrap(err)
+	}
+	certAuthority.SetRoleMap(trustedCluster.GetRoleMap())
+	if err := s.UpsertCertAuthority(certAuthority); err != nil {
```
This should still be a compare-and-swap between the existing one and the updated one.

EdwardDowling committed:
Change userca updating to a compare and swap

espadolini commented on lib/services/local/trust.go:
```diff
@@ -114,7 +114,15 @@ func (s *CA) CompareAndSwapCertAuthority(new, expected types.CertAuthority) erro
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		// might be deactivated CA
+		key = backend.Key(authoritiesPrefix, deactivatedPrefix, string(new.GetType()), new.GetName())
+		actualItem, err = s.Get(context.TODO(), key)
+		if err != nil  {
+			return trace.Wrap(err)
+		}
```
Thinking about this further, the deactivated user CA should still store the last user map, in case we downgrade 😔

espadolini replied:
The behavior of `CompareAndSwapCertAuthority` should be left alone, and it should not pretend that deactivated CAs are in some way valid.

EdwardDowling committed:
Update lib/auth/trustedcluster.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

EdwardDowling committed:
Update lib/auth/trustedcluster.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

espadolini commented on lib/services/local/trust.go:
```diff
@@ -293,6 +301,38 @@ func (s *CA) GetCertAuthorities(ctx context.Context, caType types.CertAuthType,
 }
 
+// UpdateUserCARoleMap updates the role map of the userCA of the specified existing cluster.
+func (s *CA) UpdateUserCARoleMap(ctx context.Context, existingCluster, trustedCluster types.TrustedCluster) error {
```
edit: of course it relies on a straight `Get`/`Put` for the deactivated CAs...

espadolini commented on lib/services/local/trust.go:
```diff
@@ -293,6 +301,38 @@ func (s *CA) GetCertAuthorities(ctx context.Context, caType types.CertAuthType,
 }
 
+// UpdateUserCARoleMap updates the role map of the userCA of the specified existing cluster.
+func (s *CA) UpdateUserCARoleMap(ctx context.Context, existingCluster, trustedCluster types.TrustedCluster) error {
```
```suggestion
func (s *CA) UpdateUserCARoleMap(ctx context.Context, name string, roleMap RoleMap, deactivated bool) error {
```

We don't use the `existingCluster` in any way other than to extract the name, and we don't use the `trustedCluster` other than to extract the role map - and `trustedCluster`'s name is already correct, if we found an existing trusted cluster.

espadolini commented on lib/services/local/trust.go:
```diff
@@ -293,6 +301,38 @@ func (s *CA) GetCertAuthorities(ctx context.Context, caType types.CertAuthType,
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		// CA may be deactivated.
```
We already know if the CA is supposed to be active or not, so you can have the same flow for this whole thing (using the raw `CompareAndSwap` on the backend) just by choosing between `authorities` and `authorities/deactivated` as a prefix - duplicating code sucks, but so does the way we use multiple keys to hold the state of a trusted cluster, and I really don't want deactivated CAs to become an even bigger problem than they currently are.

EdwardDowling replied:
Something like this?

espadolini reviewed (changes requested):
After talking with @fspmarshall about this, I reckon that the situation around deactivated trusted clusters won't get worse after these changes.

EdwardDowling committed:
Update lib/services/local/trust.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

EdwardDowling committed:
Remove changes to CAS and compensate in updateusercarolemap

espadolini commented on lib/auth/trustedcluster.go:
```diff
@@ -44,31 +45,44 @@ import (
+	// Update role map
+	if existingCluster != nil && !cmp.Equal(existingCluster.GetRoleMap(), trustedCluster.GetRoleMap()) {
+		if err := a.UpdateUserCARoleMap(ctx, existingCluster.GetName(), trustedCluster.GetRoleMap(),
+			existingCluster != nil && existingCluster.GetEnabled()); err != nil {
```
Already non-`nil`.

```suggestion
			existingCluster.GetEnabled()); err != nil {
```

EdwardDowling committed:
Update lib/auth/trustedcluster.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

zmb3 commented on integration/helpers/trustedclusters.go:
```diff
@@ -141,3 +142,75 @@ func WaitForActiveTunnelConnections(t *testing.T, tunnel reversetunnel.Server, c
+// TrustedClusterSetup is a grouping of configuration options describing the current trusted
+// clusters being tested used for passing info about the clusters to be tested to helper functions.
+type TrustedClusterSetup struct {
+	Ctx context.Context
```
Can we avoid storing a context like this and instead pass it as an argument to long running functions?

zmb3 commented on integration/helpers/trustedclusters.go:
```diff
@@ -141,3 +142,75 @@ func WaitForActiveTunnelConnections(t *testing.T, tunnel reversetunnel.Server, c
+	useJumpHost := tcSetup.UseJumpHost
+
+	// ensure cluster that was enabled from disabled still works
+	sshPort, _, _ := aux.StartNodeAndProxy(t, "aux-node")
```
Missing error checks here?

zmb3 commented on integration/helpers/trustedclusters.go:
```diff
@@ -141,3 +142,75 @@ func WaitForActiveTunnelConnections(t *testing.T, tunnel reversetunnel.Server, c
+	output := &bytes.Buffer{}
+	tc.Stdout = output
+	require.NoError(t, err)
+	for i := 0; i < 10; i++ {
```
Why not use `require.Eventually` here?

zmb3 commented on lib/services/local/trust.go:
```diff
@@ -293,6 +293,43 @@ func (s *CA) GetCertAuthorities(ctx context.Context, caType types.CertAuthType,
+		key = backend.Key(authoritiesPrefix, deactivatedPrefix, string(types.UserCA), name)
+	}
+
+	actualItem, err := s.Get(context.TODO(), key)
```
Why are you not using the provided context here? (same goes for below)

zmb3 commented on integration/integration_test.go:
```diff
@@ -2349,6 +2351,22 @@ func testTrustedClusters(t *testing.T, suite *integrationTestSuite) {
 }
 
+// testDisabledTrustedClusters tests creation of disabled trusted cluster
+func testDisabledTrustedClusters(t *testing.T, suite *integrationTestSuite) {
```
This test fails for me.

```
    --- FAIL: TestIntegrations/TrustedDisabledClusters (5.11s)
        trustedclusters.go:214:
            	Error Trace:	/Users/zmb/teleport/integration/trustedclusters.go:214
            	            				/Users/zmb/teleport/integration/integration_test.go:2767
            	            				/Users/zmb/teleport/integration/integration_test.go:2359
            	            				/Users/zmb/teleport/integration/integration_test.go:97
            	Error:      	Received unexpected error:
            	            	access denied to zmb connecting to 127.0.0.1 on cluster cluster-aux
            	Test:       	TestIntegrations/TrustedDisabledClusters
```

EdwardDowling replied:
I have not been able to replicate this. Is this still the case with the latest version of this branch?

r0mant replied:
This test fails for me periodically as well albeit with a different error:

```
--- FAIL: TestIntegrations (8.16s)
    --- FAIL: TestIntegrations/TrustedDisabledClusters (8.15s)
        trustedclusters.go:206:
            	Error Trace:	/Users/r0mant/code/go/src/github.com/gravitational/teleport/integration/trustedclusters.go:206
            	            				/Users/r0mant/code/go/src/github.com/gravitational/teleport/integration/integration_test.go:2771
            	            				/Users/r0mant/code/go/src/github.com/gravitational/teleport/integration/integration_test.go:2366
            	            				/Users/r0mant/code/go/src/github.com/gravitational/teleport/integration/integration_test.go:99
            	Error:      	Condition never satisfied
            	Test:       	TestIntegrations/TrustedDisabledClusters
            	Messages:   	Two clusters cannot connect to each other
```

We need to fix the flakiness before we can merge this.

EdwardDowling replied:
I think the previous max timeout for the eventually was just far too small. I have upped the amount of time it will wait before giving up

zmb3 replied:
I just pulled the latest and am still seeing the error Roman reported.

```
--- FAIL: TestIntegrations (12.38s)
    --- FAIL: TestIntegrations/TrustedDisabledClusters (12.38s)
        trustedclusters.go:206:
            	Error Trace:	/Users/zmb/teleport/integration/trustedclusters.go:206
            	            				/Users/zmb/teleport/integration/integration_test.go:2771
            	            				/Users/zmb/teleport/integration/integration_test.go:2366
            	            				/Users/zmb/teleport/integration/integration_test.go:99
            	Error:      	Condition never satisfied
            	Test:       	TestIntegrations/TrustedDisabledClusters
            	Messages:   	Two clusters cannot connect to each other
```

Additionally, it looks like the runtime of the test more than doubled, from 5 seconds to  12.

r0mant replied:
@EdwardDowling I've done more troubleshooting here and I'm pretty confident this behavior is due to caching as debug logging shows that cache removes cluster-main's user authority for some reason. Disabling caching makes the test pass with `-count 10` for me. I'm not sure there's a reliable way to rewrite this test that disables/enables authorities with our caching system. Let's disable the cache just for this particular test.

EdwardDowling replied:
I've added the lines for disabling the caching for this test.

EdwardDowling committed:
Pass ctx aroung in trustedclusterlabels integration tests

EdwardDowling committed:
Use require.Eventually to simplify trustedcluster integration test

EdwardDowling committed:
Remove superfuous error check in trusted cluster integration test

EdwardDowling committed:
Swap ctx arg around to stop linter error

r0mant commented on lib/services/trust.go:
```diff
@@ -68,4 +68,7 @@ type Trust interface {
 	DeactivateCertAuthority(id types.CertAuthID) error
+
+	// UpdateUserCARoleMap updates the role map of the userCA of the specified existing cluster.
+	UpdateUserCARoleMap(ctx context.Context, name string, roleMap types.RoleMap, deactivated bool) error
```
So is the last parameter "activated" or "deactivated"? It's "deactivated" here but "activated" in the implementation method above.

r0mant commented on lib/auth/trustedcluster.go:
```diff
@@ -69,9 +69,24 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.
+
+	// Update role map
+	if existingCluster != nil && !cmp.Equal(existingCluster.GetRoleMap(), trustedCluster.GetRoleMap()) {
+		if err := a.UpdateUserCARoleMap(ctx, existingCluster.GetName(), trustedCluster.GetRoleMap(),
```
So we've updated the role map in the CA here. What happens if anything below fails? Should we restore the original role map? Otherwise we risk the situation where it's updated in the user CA but not updated in the trusted cluster, no?

EdwardDowling replied:
I've added a solution using named returns and modifying the userca in a  defer if the err is nil.

EdwardDowling committed:
Fix arguement name in UpdateUsercarolemap

EdwardDowling committed:
Add defer to reset user ca role map if updating trustedcluster fails

EdwardDowling committed:
Disable Cache in integration test for disabled trusted cluster

r0mant reviewed (changes requested):
@EdwardDowling This fix doesn't do anything because cache policy is still set to "enabled" explicitly in GenerateConfig function: https://github.com/gravitational/teleport/blob/2847bf7e697a4f1a7e5d7e262a9aa108a1e1012d/integration/helpers/instance.go#L413

EdwardDowling committed:
Use cachpolicy passed in tconf in generateConfig

Also set enabled as the default in the default configs

EdwardDowling commented:
> @EdwardDowling This fix doesn't do anything because cache policy is still set to "enabled" explicitly in GenerateConfig function:
> 
> https://github.com/gravitational/teleport/blob/2847bf7e697a4f1a7e5d7e262a9aa108a1e1012d/integration/helpers/instance.go#L413

I've changed it so the value passed to GenerateConfig is followed and updated ApplyDefaults to set enabled by default

r0mant commented on lib/service/cfg.go:
```diff
@@ -1493,6 +1493,8 @@ func ApplyDefaults(cfg *Config) {
 	cfg.ConnectFailureC = make(chan time.Duration, 1)
 	cfg.CircuitBreakerConfig = breaker.DefaultBreakerConfig(cfg.Clock)
+
+	cfg.CachePolicy.Enabled = true
```
@EdwardDowling Any update on this? It's been 9 days.

EdwardDowling replied:
Sorry for delay have been trying to find a better way to do this that doesn't require changing the non test behavior without a load of duplicated code.

jakule replied:
@EdwardDowling in my opinion, the API is messy. `Generate()` does not generate a config, and `CreateEx()` doesn't allow starting a server with the passing config (as the passed config is modified). I think we should fix the API, but your PR is definitely not the place for that. I'm fine with creating a new method that does what `CreateEx()` promises: "creating a new instance with passed config" and using that in the test. Then we can refactor the API and update other tests. Here is a draft of what you can do https://github.com/gravitational/teleport/compare/edwarddowling/update-role-map...jakule/update-role-map-cfg?expand=1

EdwardDowling committed:
Add createWithConf to prevent cachpolicy option being overwritten

jakule commented:
@EdwardDowling Can you please rebase this PR and resolve conflicts? The current version doesn't even run GitHub checks.

EdwardDowling committed:
Swap to using new NewRole instead of NewRolev3

EdwardDowling committed:
Merge branch 'master' into edwarddowling/update-role-map

jakule commented on integration/integration_test.go:
```diff
@@ -2651,6 +2669,301 @@ func trustedClusters(t *testing.T, suite *integrationTestSuite, test trustedClus
+		tconf.Proxy.DisableWebService = false
+		tconf.Proxy.DisableWebInterface = true
+		tconf.SSH.Enabled = enableSSH
+		tconf.CachePolicy.Enabled = false
+		tconf, err := instance.GenerateConfig(t, nil, tconf)
+		require.NoError(t, err)
```
This test is still failing as the cache policy is being overridden in `GenerateConfig()`. Change the order:
```suggestion
		tconf, err := instance.GenerateConfig(t, nil, tconf)
		require.NoError(t, err)
		
		tconf.CachePolicy.Enabled = false
```

EdwardDowling committed:
Update integration/integration_test.go

Co-authored-by: Jakub Nyckowski <jakub.nyckowski@goteleport.com>

EdwardDowling committed:
Merge branch 'master' into edwarddowling/update-role-map

jakule commented on integration/helpers/instance.go:
```diff
@@ -542,12 +542,17 @@ func (i *TeleInstance) GenerateConfig(t *testing.T, trustedSecrets []*InstanceSe
+	return i.CreateWithConf(t, tconf)
+}
+
+func (i *TeleInstance) CreateWithConf(_ *testing.T, tconf *service.Config) error {
```
Can you please add a comment to this function?

EdwardDowling committed:
Merge branch 'master' into edwarddowling/update-role-map

EdwardDowling committed:
Add comment for createWithConf

EdwardDowling committed:
Merge branch 'master' into edwarddowling/update-role-map

EdwardDowling committed:
Fix formating

EdwardDowling committed:
Merge branch 'master' into edwarddowling/update-role-map
